pages/tokens.py

```diff
@@ -84,17 +95,69 @@
                 sm=6,
             ),
         ]),
-
         html.Div(id='token-table'),
-
+        #Initializes a Dash callback store named 'selected-rows' with initial data as an empty list
+        #The store is updated by Dash callbacks based on user interaction (when user makes row selection)
+        dcc.Store(id='selected-rows', data=[]),
         html.Br(),
+        #Create a confirmation dialog provider with a button(Delete Selected tokens) triggering a confirmation message.
+        #Allow users to delete tokens with users confirmation.
+        dcc.ConfirmDialogProvider(
+            children=html.Button('Delete Selected Tokens', id = 'delete-button', style = {'float': 'left',  
+            'font-size': '14px', 'width': '160px', 'display': 'block', 'margin-bottom': '10px',
+            'margin-right': '5px', 'height':'40px', 'verticalAlign': 'top', 'background-color': 'green',
+            'color': 'white',}),
+            id='confirm-delete',
+            message='Are you sure you want to delete selected token(s)?',
+        ),
         html.Button(children='Export QR codes', id='token-export', n_clicks=0, style={
             'font-size': '14px', 'width': '140px', 'display': 'block', 'margin-bottom': '10px',
             'margin-right': '5px', 'height':'40px', 'verticalAlign': 'top', 'background-color': 'green',
             'color': 'white',
         }),
     ]
 )
+generate_qrcodes_for_all_tokens() #Generate QR codes for all tokens stored in database
```

Comment:
This calls `saveAsQRCode` on every token. So if we invoke this on every load, we are creating the QR code files multiple time and overriding them over and over.
- do we need to save the QR codes as files? We need to convert them to files before downloading, but it is not clear that storing them on an arbitrary filesystem is a great idea
- is this invoked on every login or every instance launch? If on every login, see note about extraneous work above.

Reply:
@achasmita I see that you have made changes related to the second point. However, I do not see your answer to the first question.

Reply:
I was exploring other option to generate QR codes dynamically and display them in the token table while allowing user to export QR code images as files. One of the options I found was converting QR code image to a base64 encoded string. That way, we can embed image directly into HTML without the need for a separate file. This will require the following changes to the code.
  - Need to update ```saveAsQRCode``` function to generate QR codes dynamically and return the QR code image as bytes
  - Need to modify ```populate-datatable``` function to generate QR codes dynamically for each token.
  - Need to update ```export_tokens``` callback to save dynamically generated QR code images as files when exporting.